// bxssdxrk-ytdl/index.js
const { ytmp3, ytmp4 } = require('@vreden/youtube_scraper');
const yts = require('yt-search');

/**
 * Verifica se a string é uma URL válida do YouTube
 * @param {string} str - String para verificar
 * @returns {boolean}
 */
function isYouTubeURL(str) {
  if (!str || typeof str !== 'string') return false;
  
  const youtubeRegex = /^(https?:\/\/)?(www\.)?(youtube\.com\/(watch\?v=|embed\/|v\/)|youtu\.be\/)[\w-]+/i;
  return youtubeRegex.test(str);
}

/**
 * Faz download de um vídeo do YouTube em MP3.
 * @param {string} url - URL do vídeo do YouTube ou termo de busca
 * @param {string|number} [quality=128] - Qualidade do áudio (ex: 128, 192, 320)
 * @returns {Promise<object>} - Informações do vídeo + link para download
 */
async function ytdlmp3(url, quality = "128") {
  if (!url) throw new Error('URL é obrigatória.');

  // Se não for uma URL válida do YouTube, pesquisa e pega o primeiro resultado
  if (!isYouTubeURL(url)) {
    const searchResult = await yts(url);
    
    if (!searchResult.videos || searchResult.videos.length === 0) {
      throw new Error(`Nenhum vídeo encontrado para: "${url}"`);
    }
    
    url = searchResult.videos[0].url;
  }

  const result = await ytmp3(url, quality);

  if (!result.status || !result.download) {
    throw new Error(result.result || `Erro desconhecido no download MP3: ${result.message}`);
  }

  return {
    info: result.metadata,
    download: {
      status: result.status,
      link: result.download.url,
      filename: result.download.filename,
      quality: result.download.quality
    }
  };
}

/**
 * Faz download de um vídeo do YouTube em MP4.
 * @param {string} url - URL do vídeo do YouTube ou termo de busca
 * @param {string|number} [quality=360] - Qualidade do vídeo (ex: 360, 720, 1080)
 * @returns {Promise<object>} - Informações do vídeo + link para download
 */
async function ytdlmp4(url, quality = "360") {
  if (!url) throw new Error('URL é obrigatória.');

  // Se não for uma URL válida do YouTube, pesquisa e pega o primeiro resultado
  if (!isYouTubeURL(url)) {
    const searchResult = await yts(url);
    
    if (!searchResult.videos || searchResult.videos.length === 0) {
      throw new Error(`Nenhum vídeo encontrado para: "${url}"`);
    }
    
    url = searchResult.videos[0].url;
  }

  const result = await ytmp4(url, quality);
  
  if (!result.status || !result.download) {
    throw new Error(result.result || `Erro desconhecido no download MP4: ${result.message}`);
  }

  return {
    info: result.metadata,
    download: {
      status: result.status,
      link: result.download.url,
      filename: result.download.filename,
      quality: result.download.quality
    }
  };
}

/**
 * Busca vídeos no YouTube.
 * @param {string} query - Termo de busca
 * @param {number} [limit=5] - Quantidade máxima de resultados
 * @returns {Promise<object[]>} - Lista de vídeos encontrados
 */
async function ytsearch(query, limit = 5) {
  if (!query) throw new Error('Query de busca é obrigatória.');

  const result = await yts(query);
  const videos = result.videos.slice(0, limit);

  return videos.map(video => ({
    title: video.title,
    videoId: video.videoId,
    url: video.url,
    duration: video.timestamp,
    views: video.views,
    thumbnail: video.thumbnail,
    author: video.author.name
  }));
}

module.exports = {
  ytdlmp3,
  ytdlmp4,
  ytsearch
};
